src/WebManager/app/src/components/multi-tag-list.tsx

```diff
@@ -0,0 +1,329 @@
+
+                    return prevState;
+                });
+            }).catch(err => {
+                this.cancel = null;
+                this.setState({
+                    error: err.toString()
+                } as IMultiTagListState);
+
+                if (this.props.onLoadFailure) {
+                    this.props.onLoadFailure(err);
+                }
+            });
+    }
+
+    followsConvention(name: string): boolean {
+        return name.split("-").length >= 3;
+    }
+
+    makeManifest(): void {
+        this.setState({
+            multiManifestTags: []
+        } as IMultiTagListState);
+        for (let i: number = 0; i < this.state.checkedTags.length; i++)
+            this.getDigestAndSize(this.state.checkedTags[i]);
+    }
+
+    process(value: any): string {
+        if (typeof (value) === "string") {
+            try {
+                value = JSON.parse(value);
+
+            }
+            catch (err) { }
+        }
+        if (typeof (value) === "number" ||
+            typeof (value) === "string" ||
+            typeof (value) === "boolean") {
+            return this.renderPrimitive(value);
+        }
+        else if (Array.isArray(value)) {
+            return this.renderArray(value);
+        }
+        else {
+            return this.renderObject(value);
+        }
+    }
+
+    renderObject(value: any): string {
+        var cad: string = "";
+        for (let key in value) {
+            if (value.hasOwnProperty(key)) {
+                if (key == "v1Compatibility") {
+                    cad += key + ":" + this.renderJson(value[key]);
+                } else {
+                    if (key == "manifest") {
+                        cad += this.process(value[key]);
+                    }
+                    else {
+                        if (key == "content-length" || key == "docker-content-digest") {
+                            cad += key + ":" + this.process(value[key]) + ";";
+                        }
+                    }
+
+                }
+            }
+        }
+        return cad;
+    }
+
+    renderJson(value: any): string {
+        return JSON.stringify(JSON.parse(value), null, 4);
+    }
+
+
+    renderArray(value: any[]): string {
+        var cad: string = "";
+        for (let x in value) {
+            cad += this.process(x) + " ";
+        }
+        return cad
+    }
+
+    renderPrimitive(value: any): string {
+        return value.toString();
+    }
+
+    changeText(e: React.FormEvent<HTMLInputElement>) {
+        if (this.followsConvention(e.target.id) && e.target.value == "") {
+            this.hash[e.target.id] = e.target.id.split("-")[1] + "-" + e.target.id.split("-")[2];
+        } else {
+            this.hash[e.target.id] = e.target.value;
+        }
+    }
+
+    changeTag(e: React.FormEvent<HTMLInputElement>) {
+        this.setState({
+            targetTag: e.target.value.replace(/[^\x00-\x7F]/g, ""),
```

Comment:
I don't know what this is but I like it!

// seriously though, it's better to add a comment to explain what you're doing here